federation/pkg/federatedtypes/hpa.go

```diff
@@ -0,0 +1,762 @@
+	// candidate to be removed from this cluster altogether.
+	for remainingMin < 0 {
+		for _, replicas := range scheduled {
+			if replicas != nil &&
+				!((replicas.min - 1) < 0) {
+				replicas.min--
+				remainingMin++
+				if remainingMin == 0 {
+					break
+				}
+			}
+		}
+	}
+
+	return remainingMin
+}
+
+func reduceMaxReplicas(remainingMax int32, availableMaxList sets.String, scheduled map[string]*replicaNums) int32 {
+	if remainingMax < 0 {
+		// first we try reducing from those clusters which already offer max
+		if availableMaxList.Len() > 0 {
+			for _, cluster := range availableMaxList.List() {
+				replicas := scheduled[cluster]
+				if replicas != nil && !((replicas.max - replicas.min) < 0) {
+					replicas.max--
+					availableMaxList.Delete(cluster)
+					remainingMax++
+					if remainingMax == 0 {
+						break
+					}
+				}
+			}
+		}
+	}
+	// If we could not get needed replicas to reduce from already offered
+	// max above we abruptly start removing replicas from some/all clusters.
+	// Here we might make some max and min to 0, signalling that this hpa be
+	// removed from this cluster altogether
+	for remainingMax < 0 {
+		for _, replicas := range scheduled {
+			if replicas != nil &&
+				!((replicas.max - replicas.min) < 0) {
+				replicas.max--
+				remainingMax++
+				if remainingMax == 0 {
+					break
+				}
+			}
+		}
+	}
+
+	return remainingMax
+}
+
+func distributeMaxReplicas(toDistributeMax int32, lists hpaLists, rdc replicaNums,
+	currentObjs map[string]pkgruntime.Object, scheduled map[string]*replicaNums) int32 {
+	for cluster := range scheduled {
+		if toDistributeMax == 0 {
+			break
+		}
+		if scheduled[cluster] == nil {
+			continue
+		}
+		if maxReplicasNeeded(currentObjs[cluster]) {
+			scheduled[cluster].max++
+			if lists.availableMax.Len() > 0 {
+				popped, notEmpty := lists.availableMax.PopAny()
+				if notEmpty {
+					// Boundary checks have happened earlier in
+					// minReplicasReducible().
+					scheduled[popped].max--
+				}
+			}
+			// Any which ways utilise available map replicas
+			toDistributeMax--
+		}
+	}
+
+	// If we have new clusters where we can  give our replicas,
+	// then give away all our replicas to the new clusters first.
+	if lists.noHpa.Len() > 0 {
+		for toDistributeMax > 0 {
+			for _, cluster := range lists.noHpa.UnsortedList() {
+				if scheduled[cluster] == nil {
+					scheduled[cluster] = &replicaNums{min: 0, max: 0}
+				}
+				// first give away max from clusters offering them
+				// this case especially helps getting hpa into newly joining
+				// clusters.
+				if lists.availableMax.Len() > 0 {
+					popped, notEmpty := lists.availableMax.PopAny()
+					if notEmpty {
+						// Boundary checks to reduce max have happened earlier in
+						// minReplicasReducible().
+						scheduled[cluster].max++
+						scheduled[popped].max--
+						toDistributeMax--
+						continue
+					}
+				}
+				if toDistributeMax < rdc.max {
+					scheduled[cluster].max += toDistributeMax
```

Comment:
OK, but that can lead to severe imbalances (as opposed to just small rounding errors).

For example if you have 99 replicas, and 100 clusters, you will dump all 99 on the last cluster, rather than 1 in each.  Or am I missing something here?